Task 12

val-fom opened the pull request:
Hi! Please review!
- [demo](https://kottans-frontend-2018.github.io/assignments_val-fom/task_12/dist/)
- [source](https://github.com/kottans-frontend-2018/assignments_val-fom/tree/task_12/task_12/)  

Thank you!

val-fom committed:
Merge pull request #1 from val-fom/task_11

update master branch

val-fom committed:
Merge pull request #2 from kottans-frontend-2018/task_11

update

val-fom committed:
add index.html

val-fom committed:
npm init

val-fom committed:
initial setup for the project with gulp

val-fom committed:
some files renaming

val-fom committed:
some cleanup

val-fom committed:
Merge remote-tracking branch 'kottans/task_11'

val-fom committed:
Merge remote-tracking branch 'kottans/task_11'

val-fom committed:
add more markup

val-fom committed:
setInterval for clock

val-fom committed:
Force tab size in GitHub

val-fom committed:
disable autoprefixer

val-fom committed:
add gulp-sourcemaps package for Sass files

add gulp-sourcemaps package to the project to generate source maps
for the Sass to CSS compilation.

val-fom committed:
comment out unused lines

val-fom committed:
switch to scss syntax

val-fom committed:
gulpfile cleanup

val-fom committed:
prune packages

val-fom committed:
add basic layout, styles and fonts

val-fom committed:
add some styling and add responsiveness

val-fom committed:
add some styling

val-fom committed:
Merge remote-tracking branch 'kottans/master'

val-fom committed:
cut out gulp

val-fom committed:
task_12: rename app/ to src/

val-fom committed:
task_12: change folder structure for more suitable for SPA

+ some cleanup

val-fom committed:
setup webpack.config.js and add appropriate packages

val-fom committed:
continue setup webpack...

val-fom committed:
add on gh-pages

val-fom committed:
add source map

val-fom committed:
add file-loader and url-loader

val-fom committed:
implement font-awesome 4.7 from package

val-fom committed:
Merge branch 'webpack' into gh-pages

val-fom committed:
production

val-fom committed:
del dist

val-fom committed:
update README.md

viktor-yakubiv commented:
I see you just implemented WebPack and did not change nothing in the project logic.

Does it compile without issues?

viktor-yakubiv commented on .editorconfig:
```diff
@@ -0,0 +1,5 @@
+root = true
+
+[*]
+indent_style = tab
```
Most people use 2 spaces for indentation in the front-end. Another big part use 4 spaces to do it. Only few people use tabs because tabs are really unpredictable: 2 or 4 spaces in your editor, 8 spaces in Chrome and GitHub. Also, you cannot shift it easily to indent by some word or variable name if it's necessary.

Anyway, if you want to use tabs, super! Keep eyes on having all the code in the repository with the same style. Currently your `*.scss` files do not follow it.

viktor-yakubiv commented on task_12/package.json:
```diff
@@ -0,0 +1,36 @@
+			"env"
+		]
+	},
+	"author": "@val-fom",
```
Field `author` is well specified by `package.json` docs. Read about [people fields](https://docs.npmjs.com/files/package.json#people-fields-author-contributors).

viktor-yakubiv commented on task_12/src/_reset.scss:
```diff
@@ -0,0 +1,51 @@
+   License: none (public domain)
+*/
+
+html, body, div, span, applet, object, iframe, h1, h2, h3, h4, h5, h6, p, blockquote, pre, a, abbr, acronym, address, big, cite, code, del, dfn, em, img, ins, kbd, q, s, samp, small, strike, strong, sub, sup, tt, var, b, u, i, center, dl, dt, dd, ol, ul, li, fieldset, form, label, legend, table, caption, tbody, tfoot, thead, tr, th, td, article, aside, canvas, details, embed, figure, figcaption,  ...
```
For this file it's okay. For your own keep eyed on line length. It's better to do not have lines more than 80/100/120 characters long. Sometimes you want to open 2 files on screen and it creates issues.

viktor-yakubiv commented on task_12/src/_variables.scss:
```diff
@@ -0,0 +1,10 @@
+$font : 'Roboto';
+
+$green: #4F7513;
```
Pretty much good.

Commonly people use something like `$color-primary`, `$color-secondary`, `$color-warning` and so on. You may see it in the frameworks like Bootstrap. It's does not matter but that way allow you to change color scheme easily without changing meaning of variables.

viktor-yakubiv commented on task_12/src/index.html:
```diff
@@ -0,0 +1,142 @@
+		<div class="main__container">
+
+			<article class="main__item">
+				<img src="img/item__pizza_2x.png" alt="" class="item__pic">
```
You should provide heading element in each _section_, which actually can be a `<section>` or an `<article>`.

viktor-yakubiv commented on task_12/src/index.html:
```diff
@@ -0,0 +1,142 @@
+				<span class="item__price">$5.00</span>
+			</article>
+
+			<article class="main__item">
```
According to BEM, looks like class mast be like `item item--main`.

viktor-yakubiv commented on task_12/src/main.js:
```diff
@@ -0,0 +1,20 @@
+function getTime() {
+	const now = new Date();
+	let ss = now.getSeconds();
+		if (ss < 10) ss = '0' + ss;
```
Indentation may be better.

Try to avoid code repeating.

viktor-yakubiv replied:
There is good method, introduced by ES6: `String.prototype.padStart()`.

viktor-yakubiv reviewed (approved):
Pretty much good!

LiliVia reviewed (commented):
On my opinion looking very good 👍

LiliVia reviewed (approved):
approved

val-fom committed:
fix the author field in package.json

val-fom committed:
move package to devDependencies

as I'm not going to publish an npm package.
So there is no difference dep* or devDep*.

val-fom committed:
replace spaces indentation with tabs

val-fom committed:
fix line width

val-fom committed:
rename sass variables

val-fom committed:
remove unused color variable

val-fom committed:
provide heading element in each section

val-fom committed:
use String.prototype.padStart()

val-fom commented:
@viktor-yakubiv, thanks a lot for your review!!!
Yes, webpack compiles without any issues.

I've done all the fixies according to your remarks, I think)
Except two things:
1. About naming color variables. I've searched a bit and found [this discussion](https://davidwalsh.name/sass-color-variables-dont-suck#comment-247511). 
Tried to it implement in my [_variables.scss](https://github.com/kottans-frontend-2018/assignments_val-fom/blob/task_12/task_12/src/_variables.scss). What do you think about such approach?
2. BEM. I am considering `article.main__item` in this way: `main` - like a Block and  `__item` - like an element. And there is no any modifiers. Am I wrong?

viktor-yakubiv commented on task_12/src/_variables.scss:
```diff
@@ -0,0 +1,12 @@
+$font: 'Roboto';
```
Actually, when you somewhere use `font-family: $font;`, you do not know is it `serif` or `sans-serif`. You need to pass whole font list here and in the main use only variable.

Also, add fall-back fonts you prefer. When the list is `Roboto, sans-serif` and Roboto failed to load, on Windows you get Arial and on MacOS you get Helvetica. It's okay but if you prefer other fonts add it to the list.

To see current default sans-serif fonts for each system [follow the link](https://github.com/twbs/bootstrap/blob/v4-dev/scss/_variables.scss#L230).

viktor-yakubiv commented on task_12/src/main.scss:
```diff
@@ -0,0 +1,216 @@
+}
+
+body {
+	font-family: #{$font}, sans-serif;
```
It's better to use shorthand in `html` instead of splitting font configuration into 2 selectors.

viktor-yakubiv commented on task_12/src/main.scss:
```diff
@@ -0,0 +1,216 @@
+@import url('https://fonts.googleapis.com/css?family=Roboto:400,700');
+@import "reset";
+@import "variables";
```
`variables` must be first file you import because everything else depends on it.

However, first meaningful file. I mean, `@import url("Roboto")` should go before because `_variables.scss` depends on it.

Is it clear?

val-fom replied:
@viktor-yakubiv thanks again! I took into account your comments.
After completing the refactoring of my weather application I will work on
file structure of the pizza app.
Because at the moment I'm not sure how to properly structure it.

viktor-yakubiv reviewed (commented):
If you use import in `main.scss`, split styles to components and make `main.css` manifest file that imports anything related to it in the right sequence and __only__, i.e. manifest file does not contain anything except imports.

Usage of `_variables.scss` is pretty much good. I am happy you have deep understanding what and how you do. Good job!

val-fom committed:
change font-family variable name and move its declaration from body to html

val-fom committed:
move fonts imports into a separate file

val-fom committed:
split stylesheet to separate components

val-fom commented:
@viktor-yakubiv thanks again! I took into account your comments.
After completing the refactoring of my weather application I will work on
file structure of the pizza app.
Because at the moment I'm not sure how to properly structure it.
